dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -0,0 +1,173 @@
+
+    private final Predicate<C> joinCondition;
+    private final Multimap<Integer, Object[]> buffer;
+    /**
+     * Used to avoid instantiating multiple times RowN in {@link #findMatchingRows()}
+     */
+    private final ArrayRow leftRow = new ArrayRow();
+    // TODO: document the usage of those
+    private final Function<L, Integer> hashForLeft;
+    private final Function<R, Integer> hashForRight;
+    private Iterator<Object[]> leftMatchingRowsIterator;
+    private volatile boolean isKilled = false;
+
+    HashInnerJoinBatchIterator(BatchIterator<L> left,
+                               BatchIterator<R> right,
+                               ElementCombiner<L, R, C> combiner,
+                               Predicate<C> joinCondition,
+                               Function<L, Integer> hashForLeft,
+                               Function<R, Integer> hashForRight,
+                               int leftSize) {
+        super(left, right, combiner);
+        this.joinCondition = joinCondition;
+        this.hashForLeft = hashForLeft;
+        this.hashForRight = hashForRight;
+        this.buffer = LinkedListMultimap.create(leftSize);
+        this.activeIt = left;
+    }
+
+    @Override
+    public C currentElement() {
+        return combiner.currentElement();
+    }
+
+    @Override
+    public void moveToStart() {
+        left.moveToStart();
+        right.moveToStart();
+        activeIt = left;
+        buffer.clear();
+        leftMatchingRowsIterator = null;
+    }
+
+    @Override
+    public boolean moveNext() {
+        // Build HashMap from left relation
+        if (activeIt == left) {
+            while (left.moveNext()) {
+                Object[] currentRow = left.currentElement().materialize();
+                // TODO: Maybe call apply on the materialized currentRow
+                buffer.put(hashForLeft.apply(left.currentElement()), currentRow);
+            }
+            if (left.allLoaded()) {
+                activeIt = right;
+            } else {
+                return false;
+            }
+        }
+
+        // In case of multiple matches on the left side (duplicate values or hash-conflicts)
+        if (leftMatchingRowsIterator != null && findMatchingRows()) {
+            return true;
+        }
+        leftMatchingRowsIterator = null;
+        while (right.moveNext()) {
+            int rightHash = hashForRight.apply(right.currentElement());
+            Collection<Object[]> leftMatchingRows = buffer.get(rightHash);
+            if (leftMatchingRows != null) {
+                leftMatchingRowsIterator = leftMatchingRows.iterator();
+                combiner.setRight(right.currentElement());
+                if (findMatchingRows()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean findMatchingRows() {
+        while (leftMatchingRowsIterator.hasNext() && isKilled == false) {
```

Comment:
I think it's not necessary to check `isKilled` on each possible match. The kill behaviour is best effort and it's more than enough to do this once per `moveNext`

Reply:
Kind of agree. The benefit could be substantial only if we have many duplicate values or hash collisions. Will remove it.